### DISCUSSION TITLE
MediaRecorder() recording right from the beginning of the track and not waiting for the start().

### DISCUSSION THREAD

**Discussion by Hrishabh17** [aiortc/aiortc#853] [Q&A]
I have set up a peer-to-server connection and add the audio and video stream from client and once the event start is emitted, the recorder should start recording, however the recorder records the track the moment it is available. 

    class Recorder:
        def __init__(self):
            self.audio_recorder = None
            self.video_recorder = None

    async def _start_recording(self, sid, to_do, roomName):
        if to_do == 'start':
            if(self.audio_recorder is None and self.video_recorder is None):
                print(f'\nStart Recording, {roomName}\n')
                transceivers = rtcPeerConn.getTransceivers()
                receiver_tracks = [t.receiver.track for t in transceivers]
                print(receiver_tracks)

                self.audio_recorder = MediaRecorder(f"audio.mp3")
                self.video_recorder = MediaRecorder(f"video.mp4")
                if self.audio_recorder is not None and self.video_recorder is not None:
                    for track in receiver_tracks:
                        if track.kind == 'audio':
                            self.audio_recorder.addTrack(track)
                            print('inside audio track')
                        elif track.kind == 'video':
                            self.video_recorder.addTrack(track)
                            print('inside video track')

                    await self.audio_recorder.start() 
                    await self.video_recorder.start()
                    print('recording started')

        elif to_do == 'stop':
            if(self.audio_recorder is not None and self.video_recorder is not None):
                print(f'\nStop Recording, {roomName}\n')
                await asyncio.gather(self.audio_recorder.stop(), self.video_recorder.stop())
                await on_track_ended()
                print('recording stopped')
                self.audio_recorder = None
                self.video_recorder = None

    recorder = Recorder()
    
    @sio.on('recordOption')
    async def start(sid, to_do, roomName):
            await recorder._start_recording(sid, to_do, roomName)

**Comment by dmitriishepelev** [comment]
Hi! I guess I have almost [similar issue](https://github.com/aiortc/aiortc/issues/838)

**Comment by Hrishabh17** [comment]
@pydev15 is it so that the recorder adds the video/audio data to the output file after starting once again even when the recorder has stopped?

the following diagram would explain better:

1. [][][][][][][] ->{ Recorder Start }   2. [][][][][][][][][][][][][] -> {Recorder Stop} 3. [][][][][][][][][][] ->{Recorder Start} 4.[][][][][] -> {Recorder Stop} 

Let's say 1,2,3,4 are streaming audio and video tracks. Once the recorder starts and the recorder stops, the output file should contain only the 2. streaming data but it has 1 as well as 2. And if the recorder starts again, it has 3 as well as 4's data saved in the new file.

I am creating a new Instance of Media Recorder each time after stopping. 

Note: I am neither muting/removing the track nor stopping the transceiver at any time from the client side. I am just triggering the Media Recorder start and stop.

**Comment by Hrishabh17** [comment]
@jlaine could you please help us?

**Comment by jlaine** [comment]
I don't understand what you mean by "recording as soon as the track is available".

Calling the `start()` method is what causes the recorder to start pulling frames out of the received media and writing them to the output. Before you call this method, the received frames build up in a buffer, they are not lost.

**Reply by Hrishabh17** [reply]
The frames that build up in a buffer before the start() method is called are also saved in the audio/video file.
